File: lib/classifier.ts
export type DocType = 'CONTRACT' | 'SPECIFICATION' | 'BOQ' | 'DRAWING' | 'ADDENDUM';

export interface ClassificationResult {
  doc_type: DocType;
  confidence: number;
}

export function classifyDocument(text: string): ClassificationResult {
  const lowerText = text.toLowerCase();

  const keywords: Record<DocType, string[]> = {
    CONTRACT: [
      'agreement',
      'contract',
      'terms and conditions',
      'scope of work',
      'payment',
      'parties',
      'hereinafter',
    ],
    SPECIFICATION: [
      'specification',
      'specifications',
      'shall be',
      'requirements',
      'material',
      'dimensions',
      'standard',
      'quality',
    ],
    BOQ: [
      'bill of quantities',
      'boq',
      'rate',
      'unit price',
      'quantity',
      'total cost',
      'item',
      'description',
    ],
    DRAWING: [
      'drawing',
      'plan',
      'section',
      'elevation',
      'scale',
      'dimension',
      'detail',
      'architectural',
      'structural',
      'layout',
    ],
    ADDENDUM: [
      'addendum',
      'amendment',
      'modification',
      'change order',
      'revision',
      'supplement',
    ],
  };

  const scores: Record<DocType, number> = {
    CONTRACT: 0,
    SPECIFICATION: 0,
    BOQ: 0,
    DRAWING: 0,
    ADDENDUM: 0,
  };

  for (const [docType, keywordList] of Object.entries(keywords)) {
    for (const keyword of keywordList) {
      const regex = new RegExp(`\\b${keyword}\\b`, 'gi');
      const matches = lowerText.match(regex);
      scores[docType as DocType] += matches ? matches.length : 0;
    }
  }

  // Normalize scores
  const maxScore = Math.max(...Object.values(scores));
  const normalizedScores: Record<DocType, number> = {} as Record<DocType, number>;

  for (const docType in scores) {
    normalizedScores[docType as DocType] = maxScore > 0 ? scores[docType as DocType] / maxScore : 0;
  }

  // Find the document type with highest score
  let topDocType: DocType = 'CONTRACT';
  let topScore = normalizedScores.CONTRACT;

  for (const [docType, score] of Object.entries(normalizedScores)) {
    if (score > topScore) {
      topScore = score;
      topDocType = docType as DocType;
    }
  }

  // If no clear winner, default to CONTRACT with lower confidence
  if (topScore < 0.3) {
    return {
      doc_type: 'CONTRACT',
      confidence: 0.4,
    };
  }

  return {
    doc_type: topDocType,
    confidence: Math.min(topScore, 0.95), // Cap at 0.95
  };
}

File: python/app/classifier.py
"""Keyword-based document classification. Ported from lib/classifier.ts."""

from __future__ import annotations

import re

DOC_TYPES = ["CONTRACT", "SPECIFICATION", "BOQ", "DRAWING", "ADDENDUM"]

KEYWORDS: dict[str, list[str]] = {
    "CONTRACT": [
        "agreement",
        "contract",
        "terms and conditions",
        "scope of work",
        "payment",
        "parties",
        "hereinafter",
    ],
    "SPECIFICATION": [
        "specification",
        "specifications",
        "shall be",
        "requirements",
        "material",
        "dimensions",
        "standard",
        "quality",
    ],
    "BOQ": [
        "bill of quantities",
        "boq",
        "rate",
        "unit price",
        "quantity",
        "total cost",
        "item",
        "description",
    ],
    "DRAWING": [
        "drawing",
        "plan",
        "section",
        "elevation",
        "scale",
        "dimension",
        "detail",
        "architectural",
        "structural",
        "layout",
    ],
    "ADDENDUM": [
        "addendum",
        "amendment",
        "modification",
        "change order",
        "revision",
        "supplement",
    ],
}


def classify_document(text: str) -> dict:
    lower_text = text.lower()

    scores = {doc_type: 0 for doc_type in DOC_TYPES}
    for doc_type, keyword_list in KEYWORDS.items():
        for keyword in keyword_list:
            pattern = rf"\b{re.escape(keyword)}\b"
            scores[doc_type] += len(re.findall(pattern, lower_text, re.IGNORECASE))

    max_score = max(scores.values())
    normalized = {
        doc_type: (score / max_score if max_score > 0 else 0) for doc_type, score in scores.items()
    }

    top_doc_type = "CONTRACT"
    top_score = normalized["CONTRACT"]
    for doc_type, score in normalized.items():
        if score > top_score:
            top_score = score
            top_doc_type = doc_type

    if top_score < 0.3:
        return {"doc_type": "CONTRACT", "confidence": 0.4}

    return {"doc_type": top_doc_type, "confidence": min(top_score, 0.95)}
